### DISCUSSION TITLE
IsNullOrWhiteSpace with non-string

### DISCUSSION THREAD

**Discussion by jrahma** [dotnet/csharplang#5150] [Language Ideas]
Hi,

When using IsNullOrWhiteSpace with a Value returns double, for example Range control or Rating control it thriws an error:

Error CS1503: Argument 1: cannot convert from 'double' to 'string'

This should be fixed because it says Null OR and if the value is not set then it's null otherwise what is the difference between:

if (string.IsNullOrWhiteSpace(RateBusinessModel.Value))
and:

if (RateBusinessModel.Value != null)

and why do we need to convert double to string and then use IsNullOrWhiteSpace? The IsNullOrWhiteSpace should do that for us to make it easy.. the same thing applies to all other types not the double only.

[https://github.com/dotnet/maui/issues/2324#issuecomment-913199263](https://github.com/dotnet/maui/issues/2324#issuecomment-913199263)

**Comment by FaustVX** [comment]
`IsNullOrWhitespace` is in the `string` class, so this method is related to some string.
Also, a `double`, like any other `struct` type can't be null, so, the `null` check is irrelevant.

**Reply by jrahma** [reply]
but the .value can be null and that's what should be covered by IsNullOrWhitespace

**Reply by FaustVX** [reply]
* If the return type of `Value` is `double`, then `Value` **can't** be null. Maybe, its `double?`, then, you can check with :
```csharp
if (expr.Value is double value)
{ }
```
[sharplab.io](https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD5QGYAIAUAKgHQBqEANqgKa5QDOuAJgParAW0BulNAlAFgAUAG9huCbgACARmKyAnPh5VqfANzCAvsOGyAbNIBMuQsLFDJ0gMzSZh1u04B+XOVW4AvAD5cABk0hLSA)

* And again, `string.IsNullOrWhiteSpace` is related to `string` only.

**Reply by jnm2** [reply]
@FaustVX is right. `string.IsNullOrX` methods are only for strings, nothing else. For a nullable value type such as `double?`, I would recommend one of these options:

```cs
if (RateBusinessModel.Value is not null)
{
    // Use `RateBusinessModel.Value.Value`
}
```

```cs
if (RateBusinessModel.Value is { } value)
{
    // Use `value`
}
```

**Reply by jrahma** [reply]
What is the value is object type , can it have the IsNullOrWhiteSpace in the future .NET?

**Reply by HaloFour** [reply]
You can write your own helper functions or extension methods for that purpose.

```cs
namespace MyExtensions {
    public static class ObjectExtensions {
        public static bool IsNullOrWhitespace(object value) => value switch {
            null => true,
            string s => string.IsNullOrWhiteSpace(s),
            _ => false
        };
    }
}
```

Either way, this is not a language request.  API requests can be made to the [runtime](https://github.com/dotnet/runtime) repository.

**Reply by FaustVX** [reply]
Why you want to use `strig.IsNullOrWhiteSpace` with objects other than `string` ?

**Reply by jrahma** [reply]
I am just saying it will be easier if future releases of NET has one way to check for null whether it's string object or any other possible data type.

**Reply by FaustVX** [reply]
@jnm2 and I provided you a proper way to handle that using simple pattern matching.

**Reply by jnm2** [reply]
@jrahma The issue is that it's not a well-defined problem with a uniform solution that works for everybody. Unless everyone is solving the same problem the same way every time, it's best to write the checks explicitly and use your own helper methods if you end up repeating yourself too much.

For example, maybe you want `string.IsNullOrWhitespace(expr?.ToString())`. But this is arbitrary and could have unexpected behavior. It's not clear why this is a good solution for everybody or even for you.

There's no global concept of whitespace/empty/blank/unspecified/etc that works for every type. That would feel a bit unreliable and ripe for misunderstandings, like falsiness in a language such as JavaScript.
